qualtran/bloqs/data_loading/one_hot_encoding.py

```diff
@@ -0,0 +1,82 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Any, Dict, List
+
+import attrs
+import cirq
+import quimb.tensor as qtn
+from numpy._typing import NDArray
+
+from qualtran import GateWithRegisters, QAny, QUInt, Register, Side, Signature, SoquetT
+from qualtran.bloqs.basic_gates import TwoBitCSwap
+from qualtran.cirq_interop._cirq_to_bloq import _add_my_tensors_from_gate
+from qualtran.simulation.classical_sim import ClassicalValT
+
+
+@attrs.frozen
+class OneHotEncoding(GateWithRegisters):
```

Comment:
Isn't this same as the `SwapWithZero` gate we already have?

Reply:
You can essentially write a one hot encoding bloq by doing an X(q[0]) and then call `SwapWithZero` with `a` as control and `b` as target. 
The part after the first `X` gate is `SwapWithZero` bloq. 

<img width="1191" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/ce155bea-4355-4383-b7bf-4b21704e595f">


Also, `SwapWithZero` now supports N-dimensional registers so you can also do a one-hot for a multi dimensional input / output register. For example, if you have a 3d data where the input is 3 selection registers (`a_x`, `a_y`, `a_z`) and target is a 3D encoding where you would set `target[x][y][z] = 1` when `a_x = x`, `a_y = y` and `a_z = z`; you can do this again by setting `target[0][0][0] = 1` and calling a 3D swap-with-zero.

Reply:
There's no issue with swap with zero destroying the rest of the "b" register here? I've always been confused by when SwapWithZero is ok to use.

Reply:
The |B> register is assumed to be |0> in this case. The circuit after the first X gate is the same circuit as used for SwapWithZero

Reply:
cool